File: tests/test_measurement_history.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { pathToFileURL } from "node:url";
import { resolve } from "node:path";

globalThis.HTMLElement = class {
  attachShadow() {
    return this.shadowRoot = {
      innerHTML: "", addEventListener() {}, querySelector() { return null; },
      querySelectorAll() { return []; }, activeElement: null,
    };
  }
};
globalThis.customElements = {
  registry: new Map(), get(name) { return this.registry.get(name); },
  define(name, value) { this.registry.set(name, value); },
};
globalThis.window = { customCards: [] };
await import(pathToFileURL(resolve("custom_components/battery_charge_manager/frontend/battery-charge-manager.js")));
const Panel = customElements.get("battery-charge-manager-panel");

const detail = (overrides = {}) => ({
  record_type: "calibration", calibration_id: "cal", setup_id: "setup", battery_id: "battery",
  setup_revision: 1, battery_revision: 1, current_setup_revision: 2, current_battery_revision: 3,
  revision_status: "historical", used: false, usage_reason: "historical", valid: true,
  confidence: "low", quantity: 1, ports: ["A"], can_approve: true,
  net_energy_wh: 3, gross_energy_wh: 3.2, idle_energy_wh: 0.2, idle_correction_status: "applied",
  setup_snapshot: { name: "Charger" }, battery_snapshot: { name: "AA" },
  current_setup_snapshot: { name: "Charger", revision: 2 },
  current_battery_snapshot: { name: "AA", revision: 3 },
  revision_differences: [], validity_history: [], revision_approvals: [], analysis_history: [], idle_sources: [],
  chart_samples: [
    { timestamp: "2026-09-01T10:00:00Z", power_w: 3.2, net_power_w: 3, gross_energy_wh: 0, net_energy_wh: 0 },
    { timestamp: "2026-09-01T11:00:00Z", power_w: 0.2, net_power_w: 0, gross_energy_wh: 3.2, net_energy_wh: 3 },
  ],
  ...overrides,
});

function panel(admin = true) {
  const p = new Panel();
  p._hass = { language: "de", user: { is_admin: admin } };
  p._tab = "idle";
  p._state = { setups: [{ setup_id: "setup", revision: 2, port_labels: ["A"] }], batteries: [],
    selected_setup_id: "setup", session: { mode: "idle" }, active_idle_summary: {},
    idle_measurements: [], calibrations: [], selected_quantity: 1 };
  return p;
}

test("a valid low-confidence calibration is green with separate confidence and use", () => {
  const html = panel().calibrationTable([detail({ usage_reason: "lower_confidence" })], true);
  assert.match(html, /bcm-badge good[^>]*>gültig/);
  assert.match(html, /niedrig/);
  assert.match(html, /höherem Vertrauen/);
  assert.match(html, /Revision 1/);
  assert.match(html, /Details/);
  assert.match(html, /Als ungültig markieren/);
});

test("historical details and curves can be opened by non-admin users", () => {
  const p = panel(false);
  p._dialog = "measurement";
  p._measurementDetail = detail();
  const html = p.renderDialog(false);
  assert.match(html, /data-series="power"/);
  assert.match(html, /data-series="energy"/);
  assert.match(html, /Ursprüngliche Revision/);
  assert.match(html, /Aktuelle Revision/);
  assert.doesNotMatch(html, /data-action="confirm-measurement"/);
});

test("missing historical samples display a permanent absence message", () => {
  const p = panel();
  p._dialog = "measurement";
  p._measurementDetail = detail({ chart_samples: [], legacy: true });
  const html = p.renderDialog(true);
  assert.match(html, /Keine Messkurve gespeichert/);
  assert.doesNotMatch(html, /Messkurve wird aufgebaut/);
});

test("approval sends the reviewed revision pair and reason only after equivalence confirmation", async () => {
  const p = panel();
  p._dialog = "measurement";
  p._measurementDetail = detail();
  p._recordDecision = "approve";
  p._decisionReason = "Nur Beschreibung korrigiert";
  const commands = [];
  p._hass.callWS = async (payload) => {
    commands.push(payload);
    if (payload.type.endsWith("get_state")) return p._state;
    if (payload.type.endsWith("get_measurement")) return detail({ revision_status: "approved" });
    return {};
  };
  await p.handleAction("confirm-measurement");
  assert.equal(commands.length, 0);
  p._decisionConfirmed = true;
  await p.handleAction("confirm-measurement");
  assert.deepEqual(commands[0], {
    type: "battery_charge_manager/set_measurement_revision_approval", record_type: "calibration",
    record_id: "cal", approved: true, reason: "Nur Beschreibung korrigiert",
    expected_setup_revision: 2, expected_battery_revision: 3,
  });
});

test("closing a loading detail does not reopen it when the request finishes", async () => {
  const p = panel();
  let finish;
  p._hass.callWS = () => new Promise((resolve) => { finish = resolve; });
  const request = p.openMeasurement("calibration", "cal");
  await p.handleAction("close-dialog");
  finish(detail());
  await request;
  assert.equal(p._dialog, null);
});

test("a decision is blocked after the record changes while its dialog is open", async () => {
  const p = panel();
  p._dialog = "measurement";
  p._measurementDetail = detail();
  p._state.calibrations = [detail({ current_battery_revision: 4 })];
  p._recordDecision = "approve";
  p._decisionConfirmed = true;
  p._decisionReason = "Description corrected";
  const commands = [];
  p._hass.callWS = async (command) => {
    commands.push(command);
    return command.type.endsWith("get_measurement") ? detail() : p._state;
  };
  await p.confirmMeasurementDecision();
  assert.equal(commands.length, 0);
  assert.match(p._error, /Details aktualisieren/);
});

test("an in-flight subscription is disposed when the element disconnects", async () => {
  const p = panel();
  let finish;
  let unsubscribed = 0;
  p._hass.callWS = async () => p._state;
  p._hass.connection = { subscribeMessage: () => new Promise(resolve => { finish = resolve; }) };
  const connecting = p._connect();
  await Promise.resolve();
  p.disconnectedCallback();
  finish(() => { unsubscribed += 1; });
  await connecting;
  assert.equal(unsubscribed, 1);
  assert.equal(p._unsub, null);
});

test("an unstable idle baseline visibly blocks ordinary charging", () => {
  const p = panel();
  p._state.batteries = [{ battery_id:"battery", name:"AA" }];
  p._state.active_calibration_summary = { median_net_energy_wh:3, quality:"stable" };
  p._state.active_idle_summary = { usable:false, quality:"unstable", reliable_count:2 };
  const html = p.renderCharge();
  assert.match(html, /data-action="start-charge" disabled/);
  assert.match(html, /widersprüchliche Messungen prüfen/);
});

File: tests/test_measurement_history.py
from datetime import datetime, timedelta, timezone
import unittest

from test_core import (
    BatteryChargeManager, BatteryType, CalibrationRecord, ChargerSetup,
    ChargeSession, ConfigEntry, FakeHass, IdleMeasurement, MeasurementSample,
)


class MeasurementHistoryTests(unittest.IsolatedAsyncioTestCase):
    def setUp(self):
        self.manager = BatteryChargeManager(FakeHass(), ConfigEntry())
        self.setup = ChargerSetup("setup", "Charger", "switch.charger", "sensor.energy")
        self.battery = BatteryType("battery", "AA", 1700, "Li-Ion", "AA")
        self.manager.setups["setup"] = self.setup
        self.manager.batteries["battery"] = self.battery
        self.manager.selected_setup_id = "setup"
        self.manager.selected_battery_id = "battery"

    def idle(self, record_id="idle", **kwargs):
        record = IdleMeasurement(record_id, "setup", 1, self.setup.snapshot(),
                                 reliable=True, average_power_w=0.2, **kwargs)
        self.manager.idle_measurements[record_id] = record
        return record

    def calibration(self, record_id="cal", **kwargs):
        record = CalibrationRecord(
            record_id, "setup", 1, self.setup.snapshot(), "battery", 1,
            self.battery.snapshot(), 1, ["A"], net_energy_wh=3.0, **kwargs,
        )
        self.manager.calibrations[record_id] = record
        return record

    async def approve(self, kind, record_id, approved=True, reason="Description corrected"):
        await self.manager.async_set_measurement_revision_approval(
            kind, record_id, approved, reason,
            expected_setup_revision=self.setup.revision,
            expected_battery_revision=self.battery.revision if kind == "calibration" else None,
        )

    async def test_old_idle_approval_changes_use_without_rewriting_origin(self):
        record = self.idle()
        origin = record.as_dict()
        self.setup.revision = 2
        self.assertIsNone(self.manager.idle_summary("setup")["baseline_power_w"])
        await self.approve("idle", "idle")
        self.assertEqual(self.manager.idle_summary("setup")["measurement_ids"], ["idle"])
        self.assertEqual(record.setup_revision, 1)
        self.assertEqual(record.setup_snapshot, origin["setup_snapshot"])
        row = self.manager.frontend_state()["idle_measurements"][0]
        self.assertTrue(row["used"])
        self.assertEqual(row["revision_status"], "approved")
        self.assertEqual(row["current_setup_revision"], 2)
        self.setup.revision = 3
        self.assertEqual(self.manager.idle_summary("setup")["measurement_ids"], [])

    async def test_calibration_approval_is_exact_pair_and_preserves_quantity(self):
        self.calibration(confidence="high")
        self.setup.revision = 2
        self.battery.revision = 4
        await self.approve("calibration", "cal")
        self.assertEqual(self.manager.calibration_summary("setup", "battery", 1)["record_ids"], ["cal"])
        self.assertEqual(self.manager.calibration_summary("setup", "battery", 2)["record_ids"], [])
        self.battery.revision = 5
        self.assertEqual(self.manager.calibration_summary("setup", "battery", 1)["record_ids"], [])

    async def test_revoke_and_validity_are_independent_and_persist(self):
        record = self.idle()
        self.setup.revision = 2
        await self.approve("idle", "idle")
        await self.manager.async_set_measurement_validity("idle", "idle", False, "Battery attached")
        self.assertEqual(self.manager.idle_summary("setup")["measurement_ids"], [])
        await self.manager.async_set_measurement_validity("idle", "idle", True, "Reviewed")
        self.assertEqual(self.manager.idle_summary("setup")["measurement_ids"], ["idle"])
        await self.approve("idle", "idle", False, "Arrangement differs")
        self.assertTrue(record.valid)
        self.assertEqual(self.manager.idle_summary("setup")["measurement_ids"], [])
        restored = IdleMeasurement.from_dict(record.as_dict())
        self.assertEqual(len(restored.validity_history), 2)
        self.assertEqual(restored.validity_history[0]["reason"], "Battery attached")
        self.assertTrue(restored.revision_approvals[0]["revoked_at"])
        self.assertEqual(restored.revision_approvals[0]["revoke_reason"], "Arrangement differs")

    async def test_stale_or_unjustified_approval_does_not_mutate(self):
        record = self.idle()
        self.setup.revision = 2
        with self.assertRaisesRegex(Exception, "revision"):
            await self.manager.async_set_measurement_revision_approval(
                "idle", "idle", True, "Formal", expected_setup_revision=1,
            )
        with self.assertRaisesRegex(Exception, "reason"):
            await self.approve("idle", "idle", reason="  ")
        self.assertEqual(record.revision_approvals, [])
        self.manager.session = ChargeSession(mode="charging")
        with self.assertRaisesRegex(Exception, "active"):
            await self.approve("idle", "idle")

    async def test_invalid_idle_reference_excludes_dependent_calibration_until_restored(self):
        self.idle()
        self.calibration(confidence="high", idle_measurement_ids=["idle"])
        await self.manager.async_set_measurement_validity("idle", "idle", False, "Bad measurement")
        self.assertEqual(self.manager.calibration_summary("setup", "battery", 1)["record_ids"], [])
        row = self.manager.frontend_state()["calibrations"][0]
        self.assertEqual(row["usage_reason"], "invalid_idle_reference")
        self.assertTrue(row["valid"])
        await self.manager.async_set_measurement_validity("idle", "idle", True)
        self.assertEqual(self.manager.calibration_summary("setup", "battery", 1)["record_ids"], ["cal"])

    def test_usage_reports_lower_confidence_exclusion_separately_from_validity(self):
        self.calibration("low", confidence="low")
        self.calibration("high", confidence="high")
        rows = {x["calibration_id"]: x for x in self.manager.frontend_state()["calibrations"]}
        self.assertTrue(rows["high"]["used"])
        self.assertTrue(rows["low"]["valid"])
        self.assertFalse(rows["low"]["used"])
        self.assertEqual(rows["low"]["usage_reason"], "lower_confidence")

    def test_historical_details_are_on_demand_bounded_and_preserve_peak(self):
        start = datetime(2026, 9, 1, tzinfo=timezone.utc)
        samples = [MeasurementSample(
            (start + timedelta(seconds=i)).isoformat(),
            power_w=90 if i == 333 else 0.2, gross_energy_wh=i / 10000,
        ) for i in range(1500)]
        record = self.idle(samples=samples)
        self.setup.revision = 2
        self.setup.charger_model = "Corrected description"
        row = self.manager.frontend_state()["idle_measurements"][0]
        self.assertNotIn("samples", row)
        self.assertNotIn("chart_samples", row)
        detail = self.manager.measurement_details("idle", "idle")
        self.assertLessEqual(len(detail["chart_samples"]), 600)
        self.assertEqual(detail["sample_count"], 1500)
        self.assertEqual(detail["chart_samples"][0]["timestamp"], samples[0].timestamp)
        self.assertEqual(detail["chart_samples"][-1]["timestamp"], samples[-1].timestamp)
        self.assertIn(90, [s["power_w"] for s in detail["chart_samples"]])
        self.assertEqual(detail["setup_snapshot"], record.setup_snapshot)
        change = next(x for x in detail["revision_differences"] if x["field"] == "charger_model")
        self.assertEqual(change["current"], "Corrected description")

    def test_used_records_are_not_lost_behind_global_history_cutoff(self):
        self.idle(started_at="2020-01-01")
        for index in range(105):
            self.manager.idle_measurements[str(index)] = IdleMeasurement(
                str(index), "another-setup", 1, {}, started_at="2026-01-01",
            )
        self.assertIn("idle", [x["measurement_id"] for x in self.manager.frontend_state()["idle_measurements"]])

    async def test_reanalysis_retains_prior_result_and_replaces_invalid_reference(self):
        old = self.idle("old")
        old.valid = False
        self.idle("new")
        record = self.calibration(
            idle_measurement_ids=["old"], idle_baseline_power_w=0.1,
            session_started_at="2026-09-01T10:00:00+00:00",
            switch_on_at="2026-09-01T10:00:00+00:00",
            charge_finished_at="2026-09-01T11:00:00+00:00",
            samples=[
                MeasurementSample("2026-09-01T10:00:00+00:00", gross_energy_wh=0, power_w=3.2),
                MeasurementSample("2026-09-01T11:00:00+00:00", gross_energy_wh=3.2, power_w=0.2),
            ],
        )
        await self.manager.async_reanalyze_calibration(
            "cal", expected_setup_revision=1, expected_battery_revision=1,
        )
        self.assertEqual(record.idle_measurement_ids, ["new"])
        self.assertEqual(record.analysis_history[0]["idle_measurement_ids"], ["old"])
        self.assertEqual(record.analysis_revision, 2)
        self.assertEqual(record.setup_revision, 1)
        self.assertAlmostEqual(record.net_energy_wh, 3.0)
        self.assertEqual(self.manager.calibration_summary("setup", "battery", 1)["record_ids"], ["cal"])

    async def test_restoring_reliable_idle_reprocesses_pending_calibration(self):
        idle = self.idle(valid=False)
        record = self.calibration(
            idle_correction_status="pending",
            session_started_at="2026-09-01T10:00:00+00:00",
            charge_finished_at="2026-09-01T11:00:00+00:00",
            samples=[MeasurementSample("2026-09-01T11:00:00+00:00", gross_energy_wh=3.2)],
        )
        await self.manager.async_set_measurement_validity("idle", idle.measurement_id, True)
        self.assertEqual(record.idle_correction_status, "applied")
        self.assertEqual(record.idle_measurement_ids, ["idle"])
